crates/rpc/src/server/mod.rs

```diff
@@ -11,44 +10,154 @@ use crate::COMPONENT;
 
 mod api;
 
-/// Represents an initialized rpc component where the RPC connection is open, but not yet actively
-/// responding to requests.
+/// The RPC server component.
 ///
-/// Separating the connection binding from the server spawning allows the caller to connect other
-/// components to the store without resorting to sleeps or other mechanisms to spawn dependent
-/// components.
+/// On startup, binds to the provided listener and starts serving the RPC API.
+/// It connects lazily to the store and block producer components as needed.
+/// Requests will fail if the components are not available.
 pub struct Rpc {
-    api_service: api_server::ApiServer<RpcService>,
-    listener: TcpListener,
+    pub listener: TcpListener,
+    pub store: SocketAddr,
+    pub block_producer: SocketAddr,
 }
 
 impl Rpc {
-    pub async fn init(
-        listener: TcpListener,
-        store: SocketAddr,
-        block_producer: SocketAddr,
-    ) -> anyhow::Result<Self> {
-        info!(target: COMPONENT, endpoint=?listener, %store, %block_producer, "Initializing server");
-
-        let api = api::RpcService::new(store, block_producer)
-            .await
-            .context("failed to initialize RPC service")?;
+    /// Serves the RPC API.
+    ///
+    /// Note: Executes in place (i.e. not spawned) and will run indefinitely until
+    ///       a fatal error is encountered.
+    pub async fn serve(self) -> anyhow::Result<()> {
+        let api = api::RpcService::new(self.store, self.block_producer);
         let api_service = api_server::ApiServer::new(api);
 
         info!(target: COMPONENT, "Server initialized");
 
-        Ok(Self { api_service, listener })
-    }
-
-    /// Serves the RPC API.
-    ///
-    /// Note: this blocks until the server dies.
-    pub async fn serve(self) -> anyhow::Result<()> {
         tonic::transport::Server::builder()
             .accept_http1(true)
-            .add_service(tonic_web::enable(self.api_service))  // tonic_web::enable is needed to support grpc-web calls
+            // Enables gRPC-web support.
+            .add_service(tonic_web::enable(api_service))
             .serve_with_incoming(TcpListenerStream::new(self.listener))
             .await
             .context("failed to serve RPC API")
     }
 }
+
+#[cfg(test)]
+mod test {
```

Comment:
nit for the future: we usually name such modules `tests` and also put `TESTS` section separator above them.